File: usvm-python/usvm-python-main/src/main/java/org/usvm/language/SymbolForCPython.java
package org.usvm.language;

import org.jetbrains.annotations.Nullable;
import org.usvm.machine.symbolicobjects.UninterpretedSymbolicPythonObject;

public class SymbolForCPython {
    @Nullable public UninterpretedSymbolicPythonObject obj;
    public long symbolicTpCall = 0;
    public SymbolForCPython(@Nullable UninterpretedSymbolicPythonObject obj, long symbolicTpCall) {
        this.obj = obj;
        this.symbolicTpCall = symbolicTpCall;
    }

    // TODO: consider descriptor?
    @Override
    public boolean equals(Object other) {
        if (!(other instanceof SymbolForCPython))
            return false;
        if (obj == null)
            return ((SymbolForCPython) other).obj == null && symbolicTpCall == ((SymbolForCPython) other).symbolicTpCall;
        return this.obj.equals(((SymbolForCPython) other).obj) && symbolicTpCall == ((SymbolForCPython) other).symbolicTpCall;
    }

    @Override
    public int hashCode() {
        if (obj == null)
            return 0;
        return obj.hashCode();
    }
}

File: usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/interpreters/concrete/utils/SymbolForCPython.kt
package org.usvm.machine.interpreters.concrete.utils

import org.usvm.machine.symbolicobjects.UninterpretedSymbolicPythonObject

class SymbolForCPython(
    @JvmField
    val obj: UninterpretedSymbolicPythonObject?,
    @JvmField
    val symbolicTpCall: Long,
) {

    override fun equals(other: Any?): Boolean {
        if (other !is SymbolForCPython) return false
        return obj == other.obj && symbolicTpCall == other.symbolicTpCall
    }

    override fun hashCode(): Int {
        return obj.hashCode()
    }
}
